### PR TITLE
Support multiple datasources

### PR DESCRIPTION
**Pull request by tutfoo**
- added a @JooqInject annotation to reference multiple datasources
- updated ExampleApp and PostResource to use new annotation
- added unit tests
- supports backwards compatibility
- requires jdk8 for streams API

### CODE REVIEW

**Comment by tutfoo** [comment]
btw, the build is failing on the jdk7 build because the implementation uses java 8 streams. If that's not okay, I can change the implementation to work with both versions.

**Comment by saem** [comment]
I had a chance to review the code (@tutfoo is a coworker). It does break jdk7, but that's not an issue for us -- I believe Dropwizard is moving on to 8, as well.

**Comment by benjamin-bader** [comment]
Hi, JDK7 compatibility is not an issue for me, either.  I'm comfortable dropping it from the CI configuration.

I really like multi-datasource as a feature, and will be glad to have it.  The implementation is a good start.

I'm a little uncomfortable opting all users into implementing two methods of a multiple-datasource interface (its current form), regardless of whether they need it.  Features should be pay-to-play - if we don't use them, we shouldn't have to do extra work to accommodate them.

IMO, as we're dropping JDK7, I think it would be better to add a default method on `JooqConfiguration` e.g.

```java
default ImmutableList<DataSource> secondaryDataSources() {
  return ImmutableList.of(dataSourceFactory());
}
```

which would not impose extra cruft for people with single-source implementations and would still allow for customization.

Also, I'd been hoping to keep `@Context`-injection for Jooq-related objects; given Jersey 2.X, it's debatable at this point whether it's a realistic or desirable goal.  If we are going to add project-specific annotations, let's do it everywhere and keep a coherent API.

Thoughts?

**Comment by tutfoo** [comment]
This is great feedback, thanks! 

About having users implement two methods: It's implemented in a way that allows backwards compatibility, where a user *must* implement either one of the two methods:
- Override and implement getDataSourceFactory() for a single datasource solution
- Override and implement getDataSourceFactories() for a multi-datasource solution

Unfortunately, the user will not know whether to implement one over the other without reading the documentation - this is because both methods have a default implementation. 

However, I like your suggestion on using a default method to implement the secondaryDataSources(). This won't get around the must-implement one or the other of the two methods, but I could change the implementation to:
- remove the default implementation of getDataSourceFactory(), meaning the user will be forced to implement at least this method
- rename getDataSourceFactories() to getSecondaryDataSourceFactories() with a default method
- remove the dependency of getDataSourceFactory() from getSecondaryDataSourceFactories() 
- getSecondaryDataSourceFactories() will return back any additional DataSourceFactories other than the primary (defined in getDataSourceFactory())
- may require some extra logic to JooqConfiguration.run() to handle backwards compatibility

This will mean that users with single datasource solutions will notice no difference when migrating to the new version - they must override and implement getDataSourceFactory() as before. Users with multi-datasource solutions will need to override and implement getSecondaryDataSourceFactories().

What do you think?

**Comment by tutfoo** [comment]
About the @Context annotation: An early implementation we tried, used @Context @Named("slave") on the method field. However, we had to use a custom ValueFactoryProvider to extract the value from the second annotation @Named because Jersey didn't support it by default, like it does in HK2: https://hk2.java.net/hk2-api/#Injection_by_name

I raised an issue with the Jersey folks, but it was closed with reason being that @Context is JAX-RS defined and does not and will not work with HK2's @Named (works as designed): https://java.net/jira/browse/JERSEY-2937

The current implementation works with either of @JooqInject("name") or @Context for backwards compatibility. The example resources show both implementations. 

We could go with the @Named and custom ValueFactoryProvider to allow backwards compatibility as well as keeping a coherent API, though it makes the @Context annotation sort of redundant as it really doesn't really need to be there.

**Comment by benjamin-bader** [comment]
Yeah, I've bumped into the JAX-RS <-> HK2 mismatch a few times.  You're correct that a custom ValueFactoryProvider makes `@Context` redundant.  I'm OK with eliminating `@Context`, actually - far better to have a simple API than a complicated and unclear API - the back-compat burden is probably very light.

**Comment by tutfoo** [comment]
I've made some changes:
- `getDataSourceFactory()` is now _mandatory_ and defines the primary `DataSourceFactory`, similar to pre-multi-datasource-changes
- `getSecondaryDataSourceFactories()` is _optional_ and defines `DataSourceFactory`'s in addition to the primary
- `primaryDataSourceName()` is _optional_ and defines the name used to register the primary `DataSourceFactory`, `JooqFactory`, and `JooqHealthCheck`. Defaults to "jooq" if not overriden
- JDK 1.7 has been removed from the build

Please take a look and let me know if any improvements can be made.

**Comment by tutfoo** [comment]
Hi Benjamin, just checking in to see if you were able to review the recent changes.

**Comment by benjamin-bader** [comment]
Thanks for checking; I'd actually missed that you made changes at all.  I'll look at this tonight or tomorrow as I can.

File: dropwizard-jooq/src/main/java/com/bendb/dropwizard/jooq/JooqConfiguration.java [hunk 1]
```diff
@@ -1,8 +1,21 @@
...
+public interface JooqConfiguration<C extends Configuration> extends DatabaseConfiguration<C>,
+                                                                    MultiDatabaseConfiguration<C> {
+
     JooqFactory getJooqFactory(C configuration);
+
+    /**
+     * Override this method to use multiple databases, with instances referenced by name.
+     */
+    default SortedMap<String,DataSourceFactory> getSecondaryDataSourceFactories(C configuration) {
+        return new TreeMap<>();
```
Comment (benjamin-bader) [inline dropwizard-jooq/src/main/java/com/bendb/dropwizard/jooq/JooqConfiguration.java]: `return ImmutableSortedMap.of();`

**Comment by benjamin-bader** [comment]
Looks good, modulo the inline comment and the CI build failure.  The latter is simple - remove oraclejdk7 from `.travis.yml`.

**Comment by tutfoo** [comment]
Thanks for taking the time to review the changes. I've made that change to the default method (makes sense since that method needs to be overriden to be used) and rebased for a cleaner commit history when merging with yours.

**Comment by benjamin-bader** [comment]
Thanks!

Thumbs up

**Comment by tutfoo** [comment]
Hey Ben, any chance that you'll do a release anytime soon with the recent changes? Or perhaps upload a snapshot to oss sonatype?

**Pull request merged**